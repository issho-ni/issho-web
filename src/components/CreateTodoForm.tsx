import { MutationFunction } from "@apollo/react-common"
import { Form, Formik, FormikActions } from "formik"
import * as React from "react"
import { Field } from "./Field"
import { GET_TODOS } from "./TodoList"

export interface CreateTodoFormProps {
  createTodo: MutationFunction<CreateTodoResult>
}

export interface CreateTodoFormValues {
  text: string
}

export class CreateTodoForm extends React.Component<CreateTodoFormProps> {
  public render() {
    return (
      <Formik<CreateTodoFormValues>
        initialValues={{ text: "" }}
        onSubmit={this.handleCreateTodo}
      >
        {() => (
          <Form>
            <Field type="text" name="text" required />
            <button type="submit">Create Todo</button>
          </Form>
        )}
      </Formik>
    )
  }

  private handleCreateTodo = (
    variables: CreateTodoFormValues,
    actions: FormikActions<CreateTodoFormValues>
  ) =>
    this.props
      .createTodo({
        optimisticResponse: {
          __typename: "Mutation",
          createTodo: {
            __typename: "Todo",
            completedAt: null,
            createdAt: new Date(),
            id: "temp",
            ...variables,
          },
        },
        update: (proxy, { data: { createTodo } }) => {
          const data = proxy.readQuery<GetTodosResult>({ query: GET_TODOS })
          data.getTodos.push(createTodo)
          proxy.writeQuery({ query: GET_TODOS, data })
        },
        variables,
      })
      .then(() => actions.resetForm({ text: "" }))
}
