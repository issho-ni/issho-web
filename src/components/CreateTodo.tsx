import gql from "graphql-tag"
import * as React from "react"
import { Mutation } from "react-apollo"

const CreateTodoForm = React.lazy(() =>
  import("./CreateTodoForm").then(module => ({
    default: module.CreateTodoForm,
  }))
)

export const CREATE_TODO = gql`
  mutation CreateTodo($text: String!) {
    createTodo(input: { text: $text }) {
      id
      text
      createdAt
      completedAt
    }
  }
`

export class CreateTodo extends React.Component {
  public render() {
    return (
      <Mutation mutation={CREATE_TODO}>
        {createTodo => <CreateTodoForm {...{ createTodo }} />}
      </Mutation>
    )
  }
}
