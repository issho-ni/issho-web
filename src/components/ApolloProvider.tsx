import { ApolloProvider as _ApolloProvider } from "@apollo/react-common"
import { InMemoryCache, NormalizedCacheObject } from "apollo-cache-inmemory"
import ApolloClient from "apollo-client"
import { ApolloLink } from "apollo-link"
import { ErrorHandler, ErrorLink } from "apollo-link-error"
import { HttpLink } from "apollo-link-http"
import * as React from "react"
import { AuthLink } from "../lib/apollo/authLink"
import { SessionContext } from "./SessionProvider"

export interface ApolloProviderProps {
  children: React.ReactNode
}

export interface ApolloProviderState {
  client: ApolloClient<NormalizedCacheObject>
}

export class ApolloProvider extends React.Component<
  ApolloProviderProps,
  ApolloProviderState
> {
  public static contextType = SessionContext
  public context!: React.ContextType<typeof SessionContext>

  constructor(props: ApolloProviderProps) {
    super(props)

    const httpLink = new HttpLink({
      uri: GRAPHQL_ENDPOINT,
    })

    const authLink = new AuthLink(this.getToken)

    const errorLink = new ErrorLink(this.handleError)

    const client = new ApolloClient({
      cache: new InMemoryCache(),
      link: ApolloLink.from([authLink, errorLink, httpLink]),
    })

    this.state = { client }
  }

  public componentDidUpdate() {
    const { client } = this.state
    client.clearStore()
  }

  public render = () => <_ApolloProvider {...this.props} {...this.state} />

  private getToken = () => this.context.token

  private handleError: ErrorHandler = ({ graphQLErrors = [] }) => {
    for (const err of graphQLErrors) {
      switch (err.message) {
        case "ERR_AUTHENTICATION_REQUIRED":
          this.context.handleLogout()
          break
        default:
          console.error(`[GraphQL Error] ${err.message}`)
      }
    }
  }
}
