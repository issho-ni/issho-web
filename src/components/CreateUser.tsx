import gql from "graphql-tag"
import * as React from "react"
import { Mutation } from "react-apollo"

const CreateUserForm = React.lazy(() =>
  import("./CreateUserForm").then(module => ({
    default: module.CreateUserForm,
  }))
)

export const CREATE_USER = gql`
  mutation CreateUser($email: String!, $name: String!, $password: String!) {
    createUser(input: { email: $email, name: $name, password: $password }) {
      token
      user {
        email
        name
      }
    }
  }
`

export class CreateUser extends React.Component {
  public render() {
    return (
      <Mutation<CreateUserResult> mutation={CREATE_USER}>
        {createUser => <CreateUserForm {...{ createUser }} />}
      </Mutation>
    )
  }
}
