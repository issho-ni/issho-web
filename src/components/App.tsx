import * as React from "react"
import { BrowserRouter } from "react-router-dom"

const ApolloProvider = React.lazy(() =>
  import("./ApolloProvider").then(module => ({
    default: module.ApolloProvider,
  }))
)
const CreateUser = React.lazy(() =>
  import("./CreateUser").then(module => ({ default: module.CreateUser }))
)
const Dashboard = React.lazy(() =>
  import("./Dashboard").then(module => ({ default: module.Dashboard }))
)
const LoginUser = React.lazy(() =>
  import("./LoginUser").then(module => ({ default: module.LoginUser }))
)
const ProtectedRoute = React.lazy(() =>
  import("./ProtectedRoute").then(module => ({
    default: module.ProtectedRoute,
  }))
)
const SessionProvider = React.lazy(() =>
  import("./SessionProvider").then(module => ({
    default: module.SessionProvider,
  }))
)

export class App extends React.Component {
  public render() {
    return (
      <React.Suspense fallback={<h1>Loadingâ€¦</h1>}>
        <SessionProvider>
          <ApolloProvider>
            <BrowserRouter>
              <ProtectedRoute
                path="/"
                authenticated
                exact
                component={Dashboard}
              />
              <ProtectedRoute path="/join" component={CreateUser} />
              <ProtectedRoute path="/login" component={LoginUser} />
            </BrowserRouter>
          </ApolloProvider>
        </SessionProvider>
      </React.Suspense>
    )
  }
}
