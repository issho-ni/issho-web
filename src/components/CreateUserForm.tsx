import { Field, Form, Formik } from "formik"
import * as React from "react"
import { MutationFn } from "react-apollo"
import { Link } from "react-router-dom"
import { clearState } from "../utils/state"
import { SessionContext } from "./SessionContext"

export interface CreateUserFormProps {
  createUser: MutationFn<LoginUserResult>
}

export interface CreateUserFormState {
  error?: string
}

export interface CreateUserFormValues {
  email: string
  name: string
  password: string
}

export class CreateUserForm extends React.Component<
  CreateUserFormProps,
  CreateUserFormState
> {
  public static contextType = SessionContext
  public context!: React.ContextType<typeof SessionContext>
  public state: CreateUserFormState = {}

  public render() {
    const { error } = this.state

    return (
      <Formik<CreateUserFormValues>
        initialValues={{ email: "", name: "", password: "" }}
        onSubmit={this.handleCreateUser}
      >
        {() => (
          <Form>
            {error ? <span>{error}</span> : ""}
            <Field type="text" name="name" autoComplete="name" required />
            <Field type="email" name="email" autoComplete="email" required />
            <Field
              type="password"
              name="password"
              autoComplete="new-password"
              required
            />
            <button type="submit">Create Account</button>
            <Link to="/login">Log In</Link>
          </Form>
        )}
      </Formik>
    )
  }

  private handleCreateUser = (variables: CreateUserFormValues) =>
    this.props
      .createUser({ variables })
      .then(clearState(this, ["error"]))
      .then(this.context.handleLogin)
      .catch(() =>
        this.setState({
          error: "Could not create an account with that e-mail address",
        })
      )
}
