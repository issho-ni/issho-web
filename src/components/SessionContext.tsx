import * as React from "react"
import { FetchResult } from "react-apollo"
import { SessionProviderState } from "./SessionProvider"

export type SessionLoginHandler = (result: FetchResult<LoginUserResult>) => void
export type SessionLogoutHandler = () => void

export interface SessionContextState extends SessionProviderState {
  handleLogin: SessionLoginHandler
  handleLogout: SessionLogoutHandler
}

export const SessionContext = React.createContext<SessionContextState>(null)
