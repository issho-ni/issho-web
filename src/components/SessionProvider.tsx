import { InMemoryCache, NormalizedCacheObject } from "apollo-cache-inmemory"
import { ApolloClient } from "apollo-client"
import { GraphQLRequest } from "apollo-link"
import { setContext } from "apollo-link-context"
import { HttpLink } from "apollo-link-http"
import * as React from "react"
import { ApolloProvider, FetchResult } from "react-apollo"
import { SessionContext, SessionContextState } from "./SessionContext"

const LOCAL_STORAGE_SESSION_KEY = "session"

const httpLink = new HttpLink({
  uri: GRAPHQL_ENDPOINT,
})

export interface SessionProviderProps {
  children: React.ReactNode
}

export class SessionProvider extends React.Component<
  SessionProviderProps,
  SessionContextState
> {
  private client: ApolloClient<NormalizedCacheObject>

  constructor(props: SessionProviderProps) {
    super(props)

    const authLink = setContext(this.setApolloContext)

    this.client = new ApolloClient({
      cache: new InMemoryCache(),
      link: authLink.concat(httpLink),
    })

    this.state = {
      handleLogin: this.handleLogin,
      handleLogout: this.handleLogout,
      token: null,
      user: null,
    }
  }

  public componentWillMount() {
    const session = localStorage.getItem(LOCAL_STORAGE_SESSION_KEY)

    if (session) {
      const { token, user } = JSON.parse(session)
      this.setState({ token, user })
    }
  }

  public render() {
    return (
      <SessionContext.Provider value={this.state}>
        <ApolloProvider client={this.client} {...this.props} />
      </SessionContext.Provider>
    )
  }

  private handleLogin = ({ data }: FetchResult<LoginUserResult>) => {
    const { token, user } = data.loginUser
    const session = { token, user }
    localStorage.setItem(LOCAL_STORAGE_SESSION_KEY, JSON.stringify(session))
    this.setState(session)
  }

  private handleLogout = () => {
    localStorage.clear()
    this.setState({ token: null, user: null })
  }

  private setApolloContext = (_: GraphQLRequest, { headers }) => {
    const { token } = this.state

    return {
      headers: {
        ...headers,
        authorization: token ? `Bearer ${token}` : "",
      },
    }
  }
}
