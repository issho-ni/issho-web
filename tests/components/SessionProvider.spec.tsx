import * as React from "react"
import { fireEvent, render, wait, waitForElement } from "react-testing-library"
import { mocked } from "ts-jest/utils"
import { SessionContext } from "../../src/components/SessionContext"
import { SessionProvider } from "../../src/components/SessionProvider"

const data = {
  loginUser: {
    token: "not-a-valid-jwt",
    user: {
      email: "test@test",
      id: "5265b472-4b64-7014-9242-3faa40163fde",
      name: "Test",
    },
  },
}

const login = handleLogin => () => handleLogin({ data })

const component = () =>
  render(
    <SessionProvider>
      <SessionContext.Consumer>
        {({ token, user, handleLogin, handleLogout }) => (
          <div>
            Token:
            <span>{token || "none"}</span>
            <br />
            User:
            <span>{user ? user.email : "none"}</span>
            <br />
            <button onClick={login(handleLogin)}>Log in</button>
            <button onClick={handleLogout}>Log out</button>
          </div>
        )}
      </SessionContext.Consumer>
    </SessionProvider>
  )

describe("SessionProvider", () => {
  describe("with a session in local storage", () => {
    beforeEach(() => {
      mocked(localStorage.getItem).mockImplementation(() =>
        JSON.stringify(data.loginUser)
      )
    })

    afterEach(() => {
      jest.resetAllMocks()
    })

    it("sets the token and user in the state", async () => {
      const { container, getByText } = component()
      await wait(() => getByText(data.loginUser.user.email))

      expect(localStorage.getItem).toBeCalledWith("session")
      expect(container).toMatchSnapshot()
    })

    it("empties the state and storage on logout", async () => {
      const { container, getAllByText, getByText } = component()
      const button = await waitForElement(() => getByText(/log out/i))
      fireEvent.click(button)

      await wait(() => getAllByText(/none/i))

      expect(localStorage.clear).toBeCalled()
      expect(container).toMatchSnapshot()
    })
  })

  describe("with empty local storage", () => {
    it("has no token or user in the state", async () => {
      const { container, getAllByText } = component()
      await wait(() => getAllByText(/none/i))

      expect(container).toMatchSnapshot()
    })

    it("writes the token and user to the state on login", async () => {
      const { container, getByText } = component()
      const button = await waitForElement(() => getByText(/log in/i))
      fireEvent.click(button)

      await wait(() => getByText(data.loginUser.user.email))

      expect(localStorage.setItem).toBeCalledWith(
        "session",
        JSON.stringify(data.loginUser)
      )
      expect(container).toMatchSnapshot()
    })
  })
})
