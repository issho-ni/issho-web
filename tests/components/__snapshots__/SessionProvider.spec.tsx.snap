// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SessionProvider with a session in local storage empties the state and storage on logout 1`] = `
<div>
  <div>
    Token:
    <span>
      none
    </span>
    <br />
    User:
    <span>
      none
    </span>
    <br />
    <button>
      Log in
    </button>
    <button>
      Log out
    </button>
  </div>
</div>
`;

exports[`SessionProvider with a session in local storage sets the token and user in the state 1`] = `
<div>
  <div>
    Token:
    <span>
      not-a-valid-jwt
    </span>
    <br />
    User:
    <span>
      test@test
    </span>
    <br />
    <button>
      Log in
    </button>
    <button>
      Log out
    </button>
  </div>
</div>
`;

exports[`SessionProvider with empty local storage has no token or user in the state 1`] = `
<div>
  <div>
    Token:
    <span>
      none
    </span>
    <br />
    User:
    <span>
      none
    </span>
    <br />
    <button>
      Log in
    </button>
    <button>
      Log out
    </button>
  </div>
</div>
`;

exports[`SessionProvider with empty local storage writes the token and user to the state on login 1`] = `
<div>
  <div>
    Token:
    <span>
      not-a-valid-jwt
    </span>
    <br />
    User:
    <span>
      test@test
    </span>
    <br />
    <button>
      Log in
    </button>
    <button>
      Log out
    </button>
  </div>
</div>
`;
